ctrl.c

```diff
@@ -924,25 +924,35 @@ static struct pv_cmd *pv_ctrl_process_endpoint_and_reply(
 			if (pv_ctrl_process_put_file(req_fd, content_length,
 						     file_path_tmp) < 0)
 				goto out;
-			if (pv_storage_validate_file_checksum(file_path_tmp,
-							      file_name) < 0) {
+			if (pv_fs_path_exist(file_path) &&
+			    !pv_storage_validate_file_checksum(file_path,
+							       file_name)) {
+				pv_log(WARN,
+				       "object %s already exists and is valid; discarding new object upload",
+				       file_path_tmp);
+				pv_fs_path_remove(file_path_tmp, false);
+
+			} else if (pv_storage_validate_file_checksum(
+					   file_path_tmp, file_name) < 0) {
 				pv_log(WARN, "object %s has bad checksum",
 				       file_path_tmp);
 				pv_ctrl_write_error_response(
 					req_fd,
 					HTTP_STATUS_UNPROCESSABLE_ENTITY,
 					"Object has bad checksum");
 				goto out;
-			}
-			pv_log(DEBUG, "renaming %s to %s", file_path_tmp,
-			       file_path);
-			if (pv_fs_path_rename(file_path_tmp, file_path) < 0) {
-				pv_log(ERROR, "could not rename: %s",
-				       strerror(errno));
-				pv_ctrl_write_error_response(
-					req_fd, HTTP_STATUS_ERROR,
-					"Cannot rename object");
-				goto out;
+			} else {
```

Comment:
this block could go out of the else as we have a goto in the previous if and so we avoid unnecessary indentation

Reply:
the first if has no goto ... so its needed.

Reply:
first we want to send also OK response hence this else....

Reply:
maybe worth checking if we delete the _tmp file in that case? do we?